console/src/conn.rs

```diff
@@ -0,0 +1,147 @@
+use console_api::tasks::{
+    tasks_client::TasksClient, DetailsRequest, TaskDetails, TaskUpdate, TasksRequest,
+};
+use futures::stream::StreamExt;
+use std::{error::Error, pin::Pin, time::Duration};
+use tonic::{transport::Channel, Streaming};
+
+#[derive(Debug)]
+pub struct Connection {
+    target: String,
+    state: State,
+}
+
+#[derive(Debug)]
+enum State {
+    Connected {
+        client: TasksClient<Channel>,
+        stream: Streaming<TaskUpdate>,
+    },
+    Disconnected(Duration),
+}
+
+impl Connection {
+    const BACKOFF: Duration = Duration::from_millis(500);
+    pub fn new(target: String) -> Self {
+        Self {
+            target,
+            state: State::Disconnected(Duration::from_secs(0)),
+        }
+    }
+
+    async fn connect(&mut self) {
+        const MAX_BACKOFF: Duration = Duration::from_secs(5);
+
+        while let State::Disconnected(backoff) = self.state {
```

Comment:
I'd have thought we could make use of tower's `Reconnect` and a backoff layer to do this (doesn't tonic already handle reconnecting?).

Reply:
yeah, we probably could --- i wanted to be able to display to the user whether we are currently connected or reconnecting, and it felt like tracking that in a separate tower middleware would be more work than just writing a simple reconnect loop. if we end up wanting to compose more tower middleware, we can always switch later?